File: SnakeVSProj/Snake/Snake.cpp
#include "pch.h"
#include "Snake.h"


Snake::Snake(IntVector2 dir, int maxBodyPieceCount, IntVector2* snakeBody)
{
	m_SnakeBody = snakeBody;
    m_MovementDirection = dir;
    m_MaxBodyPieceCount = maxBodyPieceCount;
    m_BodyPieceCount = 1;
}

Snake::~Snake()
{

}

void Snake::SetNewDirection(IntVector2 newDirection)
{
    if (newDirection + m_SnakeBody[0] == m_SnakeBody[1])
        return;

    if (m_MovementDirection != newDirection)
    {
        m_MovementDirection = newDirection;
    }
}

bool Snake::UpdateBodyPosition()
{
    IntVector2 index = m_SnakeBody[0];
    IntVector2 newIndex = index + m_MovementDirection;

    for (int i = 2; i < m_BodyPieceCount; i++)
    {
        if (newIndex == m_SnakeBody[i])
        {
            return true;
        }
    }

    // Set new head position
    m_SnakeBody[0] = newIndex;
    // Update remaining body pieces
    for (int i = 1; i < m_BodyPieceCount; i++)
    {
        // Save current body position
        IntVector2 bodyIndex = m_SnakeBody[i];
        // Assign previous body position to the next cell
        m_SnakeBody[i] = index;
        // Pass position to the next body piece
        index = bodyIndex;
    }

    return false;
}

bool Snake::CheckForCollision(IntVector2 appleIndex)
{
    if (m_BodyPieceCount >= m_MaxBodyPieceCount)
        return false;

    return appleIndex == m_SnakeBody[0];
}

int Snake::Expand()
{
    IntVector2 lastCellIndex = IntVector2(0, 0);
    if (m_BodyPieceCount > 1)
    {
        lastCellIndex = m_SnakeBody[m_BodyPieceCount - 1];
    }
    else
    {
        lastCellIndex = m_SnakeBody[0];
    }

    lastCellIndex += m_MovementDirection * -1;

    m_SnakeBody[m_BodyPieceCount] = lastCellIndex;

    m_BodyPieceCount++;

    return m_BodyPieceCount;
}

File: SnakeUnityProj/Assets/Scripts/Game/Snake.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public class Snake : MonoBehaviour
{
    // Start is called before the first frame update
    void Start()
    {
        InitSnake();
    }

    void OnDestroy()
    {
        //Dispose of IDisposible classes here
        m_Snake.Dispose();
    }

    // Increments the snake's speed each time the snake collides with an apple and gains another body piece.
    void Update()
    {
        HandleInput();

        m_AccumulatedTime += Time.deltaTime;
        if(m_AccumulatedTime >= m_UpdateRate) {
            m_AccumulatedTime = 0.0f;
        
            StopUpdate = m_Snake.UpdateBodyPosition();

            if (StopUpdate == true)
                return;
           
            if (m_Snake.CheckForCollision(Grid.Instance.GetAppleIndex())) {
                m_BodyPieceCount = m_Snake.Expand();
                m_UpdateRate *= 0.8f;
                Grid.Instance.SpawnApple();
            }

            Grid.Instance.DrawSnake(m_SnakeBody, m_BodyPieceCount);
        }
    }

    // Check for player input to change movement direction or restart the game
    void HandleInput()
    {
        if (StopUpdate == true)
            return;

        if (Input.GetKeyDown(KeyCode.UpArrow) || Input.GetKeyDown(KeyCode.W)) {
            m_Snake.SetNewDirection(new IntVector2(0, 1));
        }
        if (Input.GetKeyDown(KeyCode.DownArrow) || Input.GetKeyDown(KeyCode.S)) {
            m_Snake.SetNewDirection(new IntVector2(0, -1));
        }
        if (Input.GetKeyDown(KeyCode.RightArrow) || Input.GetKeyDown(KeyCode.D)) {
            m_Snake.SetNewDirection(new IntVector2(1, 0));
        }
        if (Input.GetKeyDown(KeyCode.LeftArrow) || Input.GetKeyDown(KeyCode.A)) {
            m_Snake.SetNewDirection(new IntVector2(-1, 0));
        }

        if (Input.GetKeyDown(KeyCode.Space)) {
            Grid.Instance.RestartGame();
            ResetSnake();
        }

        if (Input.GetKeyDown(KeyCode.Escape))
        {
            Application.Quit();
        }
    }

    // Set up the snake object the player controls
    void InitSnake()
    {
        m_UpdateRate = 0.3f;
        m_AccumulatedTime = m_UpdateRate;

        m_BodyPieceCount = 1;
        m_MaxBodyPieceCount = 10;
        m_SnakeBody = new IntVector2[m_MaxBodyPieceCount];

        //Setting initial head position
        m_Snake = new SnakeGlue(new IntVector2(0, 0), m_MaxBodyPieceCount, m_SnakeBody);
    }

    // Destroy the old snake and start with a fresh one, and spawn in a new Apple.
    void ResetSnake()
    {
        OnDestroy();

        InitSnake();
        Grid.Instance.SpawnApple();
    }

    bool StopUpdate = false;
    float m_AccumulatedTime;
    float m_UpdateRate;
    int m_BodyPieceCount;
    int m_MaxBodyPieceCount;
    IntVector2 m_MovementDirection;
    IntVector2[] m_SnakeBody;
    SnakeGlue m_Snake;
}
